src/core/lib/event_engine/posix_engine/posix_engine.cc

```diff
@@ -482,9 +485,139 @@ EventEngine::TaskHandle PosixEventEngine::RunAfterInternal(
+                         result) mutable {
+    {
+      absl::MutexLock lock(&mu_);
+      // on_resolved called, no longer inflight.
+      GPR_ASSERT(inflight_requests_.erase(handle) == 1);
+    }
+    on_resolve(std::move(result));
+  });
+  return handle;
+}
+
+LookupTaskHandle PosixEventEngine::PosixDNSResolver::LookupSRV(
+    LookupSRVCallback on_resolve, absl::string_view name, Duration timeout) {
+  GrpcAresSRVRequest* request = new GrpcAresSRVRequest(
+      name, timeout,
+      absl::bind_front(&PosixEventEngine::PosixDNSResolver::CreateEventHandle,
+                       this),
+      event_engine_);
+  const absl::Status status =
+      request->Initialize(options_.dns_server, /*check_port=*/false);
+  if (!status.ok()) {
+    // Report back initialization failure through on_resolve.
+    on_resolve(status);
+    return LookupTaskHandle{};
+  }
+  LookupTaskHandle handle{reinterpret_cast<intptr_t>(request),
+                          aba_token_.fetch_add(1)};
+  {
+    absl::MutexLock lock(&mu_);
+    inflight_requests_.insert(handle);
+  }
+  request->Start(
+      [on_resolve = std::move(on_resolve), handle,
+       this](absl::StatusOr<std::vector<EventEngine::DNSResolver::SRVRecord>>
+                 result) mutable {
+        {
+          absl::MutexLock lock(&mu_);
+          // on_resolved called, no longer inflight.
+          GPR_ASSERT(inflight_requests_.erase(handle) == 1);
+        }
+        on_resolve(std::move(result));
+      });
+  return handle;
+}
+
+LookupTaskHandle PosixEventEngine::PosixDNSResolver::LookupTXT(
+    LookupTXTCallback on_resolve, absl::string_view name, Duration timeout) {
+  GrpcAresTXTRequest* request = new GrpcAresTXTRequest(
+      name, timeout,
+      absl::bind_front(&PosixEventEngine::PosixDNSResolver::CreateEventHandle,
+                       this),
+      event_engine_);
+  const absl::Status status =
+      request->Initialize(options_.dns_server, /*check_port=*/false);
+  if (!status.ok()) {
+    // Report back initialization failure through on_resolve.
+    on_resolve(status);
+    return LookupTaskHandle{};
+  }
+  LookupTaskHandle handle{reinterpret_cast<intptr_t>(request),
+                          aba_token_.fetch_add(1)};
+  {
+    absl::MutexLock lock(&mu_);
+    inflight_requests_.insert(handle);
+  }
+  request->Start([on_resolve = std::move(on_resolve), handle,
+                  this](absl::StatusOr<std::string> result) mutable {
+    {
+      absl::MutexLock lock(&mu_);
+      // on_resolved called, no longer inflight.
+      GPR_ASSERT(inflight_requests_.erase(handle) == 1);
+    }
+    on_resolve(std::move(result));
+  });
+  return handle;
+}
+
+bool PosixEventEngine::PosixDNSResolver::CancelLookup(LookupTaskHandle handle) {
+  absl::MutexLock lock(&mu_);
+  auto iter = inflight_requests_.find(handle);
+  if (iter != inflight_requests_.end()) {
+    reinterpret_cast<GrpcAresRequest*>(iter->keys[0])->Cancel();
+    return true;
```

Comment:
You'll want to remove the handle from `inflight_requests_` if it was found. Otherwise, you risk the `GrpcAresRequest*` pointer being garbage. Consider `flat_hash_set::extract`.

Reply:
That sounds good to me.